// control-systems-js/dist/cjs/lib/impulse/impulse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Impulse = void 0;
const complex_1 = require("../math/complex");
class Impulse {
    constructor(calculateTransferFunction, inverseLaplace, convolution) {
        this.calculateTransferFunction = calculateTransferFunction;
        this.inverseLaplace = inverseLaplace;
        this.convolution = convolution;
    }
    calculatePoints(tf, timeRange) {
        // impulse = delta(t)
        const impulseExpression = {
            numerator: [(0, complex_1.complex)(1, 0)],
            denominator: [(0, complex_1.complex)(1, 0)],
        };
        // impulse response = tf * delta(t)
        const responseFunction = this.convolution.execute(tf, impulseExpression);
        const points = timeRange.map((time) => {
            const timeResponse = this.inverseLaplace.execute((s) => this.calculateTransferFunction.calculateValue(responseFunction, s), time);
            return { x: time, y: timeResponse };
        });
        return points;
    }
}
exports.Impulse = Impulse;
