// control-systems-js/dist/cjs/lib/step/step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Step = void 0;
const complex_1 = require("../math/complex");
class Step {
    constructor(calculateTransferFunction, inverseLaplace, convolution) {
        this.calculateTransferFunction = calculateTransferFunction;
        this.inverseLaplace = inverseLaplace;
        this.convolution = convolution;
    }
    calculatePoints(tf, timeRange) {
        // step = 1/s
        const stepExpression = {
            numerator: [(0, complex_1.complex)(1, 0)],
            denominator: [(0, complex_1.complex)(1, 0), (0, complex_1.complex)(0, 0)],
        };
        // step response = tf * 1/s
        const responseFunction = this.convolution.execute(tf, stepExpression);
        const points = timeRange.map((time) => {
            const timeResponse = this.inverseLaplace.execute((s) => this.calculateTransferFunction.calculateValue(responseFunction, s), time);
            return { x: time, y: timeResponse };
        });
        return points;
    }
}
exports.Step = Step;
